drafts/zip-daira-mempool-limitation.rst

```diff
@@ -0,0 +1,192 @@
+  Category: Network
+  Created: 2019-09-09
+  License: MIT
+
+
+Terminology
+===========
+
+The key words "MUST", "SHOULD", and "MAY" in this document are to be interpreted
+as described in RFC 2119. [#RFC2119]_
+
+
+Abstract
+========
+
+This proposal specifies a change to the behaviour of zcashd nodes intended to
+mitigate denial-of-service from transaction flooding.
+
+
+Motivation
+==========
+
+Adoption of this proposal would increase robustness of Zcash nodes against
+denial-of-service attack, in particular attacks that attempt to exhaust node
+memory.
+
+Bitcoin Core added size limitation for the mempool in version 0.12
+[#BitcoinCore-PR6722]_, defaulting to 300 MB. This was after Zcash forked from
+Bitcoin Core.
+
+
+Requirements
+============
+
+The memory usage of a node’s mempool should be bounded.
+
+The eviction policy should as far as possible not be “gameable” by an adversary,
+i.e. an adversary should not be able to cause legitimate transactions (that do not
+themselves present any denial-of-service problem) to be preferentially evicted
+relative to its own transactions.
+
+Any configuration options should have reasonable defaults, i.e. without changing
+relevant configuration, a node should be adequately protected from denial-of-service
+via mempool memory exhaustion.
+
+
+Non-requirements
+================
+
+The current architecture of Zcash imposes fundamental limits on scaling of
+transaction throughput. This proposal does not increase the aggregate transaction
+capacity of the network. (The Blossom upgrade does increase transaction capacity,
+by a factor of two [#zip-0208]_.)
+
+Denial-of-service issues in the messaging layer of the peer-to-peer protocol are
+out of scope for this proposal.
+
+This proposal is focused primarily on memory exhaustion attacks. It does not
+attempt to use fees to make denial-of-service economically prohibitive, since that
+is unlikely to be feasible while maintaining low fees for legitimate users. It
+does not preclude changes in fee policy.
+
+
+Specification
+=============
+
+This specification describes the intended behaviour of zcashd nodes. Other node
+implementations MAY implement the same or similar behaviour, but this is not a
+requirement of the network protocol. Thus, RFC 2119 conformance keywords below are
+to be interpreted only as placing requirements on the zcashd implementation (and
+potentially other implementations that have adopted this specification in full).
+
+The mempool of a node holds a set of transactions. Each transaction has a cost,
+which is an integer defined as:
+
+  max(serialized transaction size in bytes, 4000) + fee_penalty
+
+where ``fee_penalty`` is 16000 if the transaction pays a fee less than
+10000 zatoshi, otherwise 0.
+
+Each node also MUST hold a FIFO queue RecentlyEvicted of pairs (txid, time), where
+the time indicates when the given txid was evicted. This SHOULD be empty on node
+startup. The size of RecentlyEvicted MUST never exceed 10000 entries.
```

Comment:
```suggestion
startup. The size of RecentlyEvicted SHOULD never exceed 100000 entries.
```

Reply:
Other node implementations might manage this queue in ways that do not cause memory pressure.

Reply:
Hmm I think maybe it does - so is the maximum effective size of the RecentlyEvicted list is the nominal size divided by the number of equivalent transactions that can be created using the same funds but different transaction IDs? Without the attacker risking the transactions being mined?

Reply:
Is this now explained adequately in the rationale section?